### ISSUE TITLE
Slurm does not send soft kill message

### ISSUE THREAD

**Issue by pditommaso** [nextflow-io/nextflow#2163]
Slurm cluster by default do not send a notification message prior to killing a job. 

When this happens the job wrapper created by NF terminates abruptly without saving the .exitcode file and therefore nextflow cannot recognise the job termination.  

This could be solved by adding the following signal on the job submission, eg 

```
--signal=USR2
```

See also:

http://github.com/nextflow-io/nextflow/blob/6aa102430972266c4e19663926e68c49d1c41c80/modules/nextflow/src/main/resources/nextflow/executor/command-run.txt#L107-L107

Read more: 
https://slurm.schedmd.com/sbatch.html#OPT_signal

**Comment by phue** [comment]
That's interesting..
I've been running into this issue before, but I kept thinking it was a singularity problem related to https://github.com/nextflow-io/nextflow/issues/1561

Interestingly, the problem only occurs when a job gets killed because of walltime limit, but not when it runs oom

**Comment by pditommaso** [comment]
You may want to try adding `clusterOptions = '--signal=USR2'` in your config

**Comment by phue** [comment]
I took a closer look into this, and it seems my problem was only partially related.

It probably depends on the local slurm configuration but running nextflow w/o singularity did not require this setting for me.
However, when running with singularity and setting `clusterOptions = '--signal=USR2'` , it seems that also singularity itself does not properly propagate soft kill signal(s) leading to the same symptom: there is no .exitcode written and the job is in a zombie state until the next queue polling detects that it is not running anymore.
As mentioned before, the problem also happens when a job hits the walltime limit and slurm sends `SIGKILL` to the job, but it does not happen when the job runs out-of-memory, presumably because slurm sends `SIGTERM` in that case, which seems to be handled differently.
I could get rid of the problem by explicitly sending the signal to the shell process of the job using `'--signal=B:USR2'`, in which case it seems to be propagated correctly.

I believe my problem is not limited to the slurm executor however but https://github.com/nextflow-io/nextflow/issues/1561 appears to be caused by the same singularity quirk and was reported for LSF originally

TL;DR
The combination of slurm and singularity requires `'--signal=B:USR2'` in order to work as expected

**Comment by pditommaso** [comment]
Thanks a lot, Patrick for this feedback this is very useful. I guess default change across different slurm deployment therefore may not be always needed. 

Not sure to understand however the need for `B:`, the doc says: 

> Use the "B:" option to signal only the batch shell, none of the other processes will be signaled

do you think it means the when using this option the USR2 signal is only sent to the job root process?

**Comment by phue** [comment]
I do not fully understand why this is needed either. My best guess is that when `B:` is omitted, the signal might only be propagated to child processes of the container, but not the singularity process itself.

**Comment by pditommaso** [comment]
Reading [this](https://bugs.schedmd.com/show_bug.cgi?id=927) I think that *processes* mentioned in the docs should be read as slurm jobs launcher by the parent (batch) job. Indeed it says 

> By default all job steps will be signaled, but not the batch shell itself

instead 

> Use the "B:" option to signal only the batch shell, none of the other processes will be signaled
  
This should be consistent with your observation.

**Comment by pditommaso** [comment]
When turned into a directive, do you think the following syntax works? 

```
#SBATCH --signal B:USR2
``

I mean without the `=` symbol

**Comment by phue** [comment]
> When turned into a directive, do you think the following syntax works?
> 
> ```
> #SBATCH --signal B:USR2
> ```
> 
> I mean without the `=` symbol

Just tested that, works fine (slurm v20.11.5)

**Comment by pditommaso** [comment]
Thanks a lot! preparing a patch soon

**Comment by pditommaso** [comment]
Solved

**Comment by pditommaso** [comment]
Included in release `21.06.0-edge`.
